src/app/zeko/da_layer/lib/client.ml

```diff
@@ -30,9 +32,137 @@ let dispatch ?(max_tries = 5) ~logger
 let post_batch ~logger ~node_location ~ledger_openings ~batch =
   dispatch ~logger node_location Rpc.Post_batch.v1 { ledger_openings; batch }
 
-let get_batch ~logger ~node_location ~ledger_hash :
+let query_batch ~logger ~node_location ~ledger_hash :
     (Batch.t option, Error.t) result Deferred.t =
   dispatch ~logger node_location Rpc.Get_batch.v1 ledger_hash
 
-let get_all_keys ~logger ~node_location () =
+let query_all_keys ~logger ~node_location () =
   dispatch ~logger node_location Rpc.Get_all_keys.v1 ()
+
+let query_batch_source ~logger ~node_location ~ledger_hash =
+  dispatch ~logger node_location Rpc.Get_batch_source.v1 ledger_hash
+
+module Config = struct
+  type t = { nodes : Host_and_port.t Cli_lib.Flag.Types.with_name list }
+  [@@deriving fields]
+
+  let of_string_list uris =
+    { nodes =
+        List.mapi uris ~f:(fun i s ->
+            Cli_lib.Flag.Types.
+              { value = Host_and_port.of_string s
+              ; name = sprintf "da-node-%d" i
+              } )
+    }
+end
+
+let distribute_batch ~logger ~config ~ledger_openings ~batch ~quorum =
+  let%bind signatures =
+    Deferred.List.map ~how:`Parallel (Config.nodes config)
+      ~f:(fun node_location ->
+        post_batch ~logger ~node_location ~ledger_openings ~batch )
+    |> Deferred.map ~f:(List.filter_map ~f:Result.ok)
+  in
+  if List.length signatures >= quorum then return (Ok signatures)
+  else return (Error (Error.of_string "Quorum not reached"))
+
+module Sequencer = struct
+  type t =
+    { logger : Logger.t
+    ; config : Config.t
+    ; quorum : int
+    ; q : unit Async.Sequencer.t
+    ; mutable signatures : Signature.t list Deferred.t Ledger_hash.Map.t
+    ; mutable last_distributed_batch : Ledger_hash.t option
+    }
+
+  let create ~logger ~config ~quorum =
+    { logger
+    ; config
+    ; quorum
+    ; q = Async.Sequencer.create ~continue_on_error:false ()
+    ; signatures = Ledger_hash.Map.empty
+    ; last_distributed_batch = None
+    }
+
+  let distribute_batch t ~ledger_openings ~batch =
```

Comment:
What does this do?

Reply:
I've changed it to `enqueue_distribute_batch` in later commit. `distribute_batch` sends the batch to all the da nodes in the config. Enqueueing is to ensure strict order.